### PR TITLE
wayland: Manually implement wl_shm/wl_shm_pool

### PR DESCRIPTION
**Pull request by RAOF**
The implementation in `libwayland-server` has a number of annoying flaws, including being thread-unsafe, having awkward interactions with `SIGBUS` handling, and (least importantly) crashing the server in some client error circumstances.

Because we have fewer constraints an a richer implementation environment, it is reasonable to implement `wl_shm` and `wl_shm_pool` ourselves, making the surrounding code simpler.

Closes: #2669

### CODE REVIEW

File: src/server/frontend_wayland/shm.cpp [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        std::shared_ptr<mir::shm::RWMappableRange> data,
+        mir::geometry::Size size,
+        mir::geometry::Stride stride,
+        MirPixelFormat format)
+    : weak_buffer{buffer},
+      wayland_executor{std::move(wayland_executor)},
+      data{std::move(data)},
+      size_{std::move(size)},
+      stride_{stride},
+      format_{format}
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/shm.cpp:146]: Not our usual indentation

**Review by AlanGriffiths — APPROVED** [review]
One nit, but we presumably want the prerequisite landed first

bors delegate

File: src/server/frontend_wayland/shm.cpp (hunk 1 above)
Comment (Saviq) [reply src/server/frontend_wayland/shm.cpp]: @AlanGriffiths I've fixed, is that what you had in mind? 7631fb1

**Review by Saviq — CHANGES_REQUESTED** [review]
This still fails like this on some occasions:

```plain
 ./tests/test_bad_buffer.cpp:101: Failure
Failed
Expected protocol error not raised
[2022-12-16 12:48:48.544331] <information> mirserver: Stopping
[  FAILED  ] BadBufferTest.test_truncated_shm_file (17 ms)
```

**Comment by RAOF** [comment]
This has to be a race of some kind, but I can't see it and I can't seem to reproduce locally.

**Comment by Saviq** [comment]
@RAOF I'm able to reproduce this 100% in LXD/schroot here. I've modified my environment a bit to help:

```diff
diff --git a/spread.yaml b/spread.yaml
index 0cbcf89721..0bf620c964 100644
--- a/spread.yaml
+++ b/spread.yaml
@@ -76,3 +76,5 @@ path:
 
 exclude:
     - .git
+    - build-*
+    - install-*
diff --git a/spread/build/sbuild/task.yaml b/spread/build/sbuild/task.yaml
index 8606a6c889..8173f46ea7 100644
--- a/spread/build/sbuild/task.yaml
+++ b/spread/build/sbuild/task.yaml
@@ -49,6 +49,7 @@ execute: |
       --dist=${RELEASE}
       --build=${DEB_BUILD_ARCH}
       --host=${DEB_HOST_ARCH}
+      --finished-build-commands='sleep 3600'
     )
 
     SCHROOT_NAME="${RELEASE}-${DEB_BUILD_ARCH}"
@@ -83,7 +84,8 @@ execute: |
       ln -vs "${DEBOOTSTRAP_SCRIPTS}/gutsy" "${DEBOOTSTRAP_SCRIPTS}/${RELEASE}"
     fi
 
-    sg sbuild -c "${MKSBUILD_OPTS[*]}"
+    sg sbuild -c "${MKSBUILD_OPTS[*]}" || /bin/true
+    sed -i 's/aufs/none/' /etc/schroot/chroot.d/sbuild-*
 
     echo "export XDG_RUNTIME_DIR=/tmp" >> debian/opts.mk
```

Then, letting spread go:
```sh
$ spread-mir-ci.spread lxd:ubuntu-22.04:spread/build/sbuild:ubuntu_devel
```

After it builds, in another terminal:
```sh
$ lxc shell spread<TAB>
lxd> $ schroot --run --chroot $( schroot --list --all-sessions )
chroot> $ cd /build/mir-<TAB>/mir-2.10.0/build-amd64/
chroot> $ export XDG_RUNTIME_DIR=/tmp
chroot> $ /usr/libexec/wlcs/wlcs lib/miral_wlcs_integration.so --gtest_filter=BadBufferTest.test_truncated_shm_file
Note: Google Test filter = BadBufferTest.test_truncated_shm_file
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from BadBufferTest
[ RUN      ] BadBufferTest.test_truncated_shm_file
./tests/test_bad_buffer.cpp:101: Failure
Failed
Expected protocol error not raised
[  FAILED  ] BadBufferTest.test_truncated_shm_file (10 ms)
[----------] 1 test from BadBufferTest (10 ms total)

[----------] Global test environment tear-down
[==========] 1 tests from 1 test cases run. (10ms total elapsed)
[  PASSED  ] 0 tests
[  FAILED  ] 1 test failed:
[  FAILED  ] BadBufferTest.test_truncated_shm_file
```

**Comment by RAOF** [comment]
Aha! Apparently, building using a ZFS-backed LXD fails for me, but on my system with a dir-backed LXD it works. :woman_shrugging: 

And now, I can reproduce, and come to the horrifying revalation: it only fails when built with optimisations.

**Review by Saviq — APPROVED** [review]
Go!

bors r=AlanGriffiths

**Pull request merged**